src/placementEngine.js

```diff
@@ -28,21 +21,47 @@ const incrementVisits = update('visits', inc)
  * @param {Object} custom The custom state object.
  * @returns {Promise} A promise that resolves to the promos.
  */
-function placementEngine (storage, promos, custom) {
+function placementEngine (storage, promos, custom = {}) {
   // Generate a seed value for the PRNG.
   const seed = Date.now()
 
+  // Load the user state.
   const user = getUser(storage)
 
-  // The pipeline contains the steps in the placement engine algorithm.
-  const pipeline = pipe(
-    updateUser(storage, incrementVisits),
-    createContext(custom),
-    placePromos(seed, promos),
-    resolvePromos(storage)
-  )
+  // Create the bus signal.
+  const bus = new Bus()
+
+  // A function that emits a `click` event on the bus.
+  const onClick = promo => bus.next({ type: 'click', promo })
+
+  // A function that emits a `close` event on the bus.
+  const onClose = promo => bus.next({ type: 'close', promo })
+
+  // A function that emits a `view` event on the bus.
+  const onView = promo => bus.next({ type: 'view', promo })
+
+  // A function that emits a `refresh` event on the bus.
+  const onRefresh = promo => bus.next({ type: 'refresh', promo })
+
+  // Create the initial state object.
+  const initialState = { seed, user }
+
+  // The stateful signal emits the current placement engine state whenever an
+  // event is emitted on the bus.
+  const statefulSignal = bus
+    // Throttle events to a maximum rate of one per second.
+    .throttle(1000)
```

Comment:
can you talk about throttling more?

So this is throttle events which the client emits? like scroll events? but curious does this have an impact on the other events like a click event?

I might have completely miss-understood this.. 😬

Reply:
Sure 😄 

Throttling rate-limits the events. If more than one event is received within a second, then the _first_ one received within that window will be emitted, the rest will be dropped.

This is mainly to avoid the scroll events causing too much re-processing of the promos.

Reply:
So it has to be the exact same time of event to be throttled?
Like what happens if there was a scroll and a click in the same period of time?

Reply:
Oh, good point. That would be a problem.

I guess we could just do the throttling on the [TC side](https://github.com/conversation/tc/pull/9150/files#diff-a61f07464842430dbfc819e6a93e31deR59) 🤔

Reply:
i.e. Throttle the raw scroll events somehow.

Reply:
Yeah, I think that makes sense to me.

Reply:
Cool, I will make that change 👍